src/python/pants/engine/internals/graph.py

```diff
@@ -98,85 +98,88 @@ async def resolve_unexpanded_targets(addresses: Addresses) -> UnexpandedTargets:
-    if build_target.has_field(Sources):
-        sources_field = build_target.get(Sources)
-        supports_subtargets = not sources_field.indivisible
-    else:
-        supports_subtargets = False
-
-    if not build_target.has_field(Dependencies):
-        supports_subtargets = False
-
-    if not supports_subtargets:
-        # If a target type does not support dependencies or has no Sources field (or the Sources field is
-        # indivisible), we do not split it, as that would prevent the BUILD target from depending on its splits.
-        return Subtargets(build_target, ())
-
-    # Generate a subtarget per source.
-    paths = await Get(SourcesPaths, SourcesPathsRequest(build_target[Sources]))
-    wrapped_subtargets = await MultiGet(
-        Get(
-            WrappedTarget,
-            Address,
-            generate_subtarget_address(address, full_file_name=subtarget_file),
-        )
-        for subtarget_file in paths.files
+def target_types_to_generate_targets_requests(
+    union_membership: UnionMembership,
+) -> TargetTypesToGenerateTargetsRequests:
+    return TargetTypesToGenerateTargetsRequests(
+        {
+            request_cls.generate_from: request_cls  # type: ignore[misc]
+            for request_cls in union_membership.get(GenerateTargetsRequest)
+        }
     )
-    return Subtargets(build_target, tuple(wt.target for wt in wrapped_subtargets))
 
 
 @rule
 async def resolve_target(
     address: Address,
     registered_target_types: RegisteredTargetTypes,
     union_membership: UnionMembership,
+    target_types_to_generate_requests: TargetTypesToGenerateTargetsRequests,
 ) -> WrappedTarget:
     if address.is_file_target:
-        build_target = await Get(WrappedTarget, Address, address.maybe_convert_to_build_target())
-        subtarget = generate_subtarget(
-            build_target.target, full_file_name=address.filename, union_membership=union_membership
+        wrapped_generator_tgt = await Get(
+            WrappedTarget, Address, address.maybe_convert_to_build_target()
         )
-        return WrappedTarget(subtarget)
+        generator_tgt = wrapped_generator_tgt.target
+        # Not all languages support file targets (Go and Terraform). If we got a file address, use
+        # that as an alias for the BUILD target.
+        #
+        # This code path will go away once we remove file addresses in favor of generated target
+        # addresses, e.g. path/to:generator:generated, because we will error in that case when
+        # there is no generator for the address.
+        if not target_types_to_generate_requests.is_generator(generator_tgt):
+            return wrapped_generator_tgt
+        generate_request = target_types_to_generate_requests[type(generator_tgt)]
+        generated = await Get(
+            GeneratedTargets, GenerateTargetsRequest, generate_request(generator_tgt)
+        )
+        requested = next((tgt for tgt in generated if tgt.address == address), None)
```

Comment:
This linear scan isn't ideal... `GeneratedTargets` should maybe have a dict inside?

As to the API, are we requiring a relationship between the generator and generated that means that this only needs to compare the `name` (or filename for now, I suppose)? Would be good to tighten that up so that you cannot generate targets "above" yourself in the filesystem (which file `Addresses` ban via their parsing, but which will take some explicit work here).

Reply:
Great idea! Changed to `FrozenDict[Address, Target]`.

> As to the API, are we requiring a relationship between the generator and generated that means that this only needs to compare the name (or filename for now, I suppose)?

I think the only notion of "above" and "below" (i.e. locality) could be the `sources` field of the generated target. Generated targets have no concrete physical location, only the location of their generator + some `generator_name`. If you generate `project:lib#subdir/foo.py`, you're not actually creating a target in `project/subdir`. You're creating a target with `spec_path=project/` and `sources=["subdir/foo.py"]`. It's not possible to create a target in `project/subdir` - the `spec_path` is always the `target_generator`. So, I don't think we have to worry about generating targets above you. Does this make sense?

*We could add eager validation that the `Address`es of `GeneratedTargets` aren't cheating.

Reply:
Cool, got a small followup patch to validate that the generated addresses are well-formed.

Reply:
>  If you generate `project:lib#subdir/foo.py`, you're not actually creating a target in `project/subdir`. You're creating a target with `spec_path=project/` and `sources=["subdir/foo.py"]`. It's not possible to create a target in `project/subdir` - the `spec_path` is always the `target_generator`. So, I don't think we have to worry about generating targets above you. Does this make sense?

Mostly, yea. I just want to bias in the direction of being restrictive (in whichever dimensions) and then only loosening once we know the coast is clear.